File: app/src/main/java/yagoc/board/BoardRules.java
package yagoc.board;

import yagoc.pieces.PieceColor;

import java.util.Collection;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static yagoc.Yagoc.logger;
import static yagoc.pieces.PieceType.King;
import static yagoc.pieces.Pieces.none;

public class BoardRules {
    public static boolean isCorrectMove(BoardReader board, Square from, Square to) {
        return isCorrectMove(board, new Move(board.pieceAt(from), from, to));
    }

    public static boolean isCorrectMove(BoardReader board, Move move) {
        return move.fromPiece().isCorrectMove(board, move);
    }

    static boolean cannotMoveWithoutBeingCheck(BoardReader board) {
        return generateMoves(board).stream().noneMatch((move) -> moveDoesNotCreateCheck(board, move));
    }

    public static boolean isInCheck(BoardReader board, PieceColor color) {
        Square kingSquare = Square.allSquares.stream()
                .filter((square) -> board.pieceAt(square).pieceType() == King && board.pieceAt(square).color() == color)
                .findAny().orElseThrow(() -> new RuntimeException("Could not find " + color + " king!"));

        return Square.allSquares.stream()
                .anyMatch((from) -> !board.pieceAt(from).equals(none) && !board.pieceAt(from).color().equals(color) && isCorrectMove(board, from, kingSquare));
    }

    public static boolean isCheckmate(BoardReader board) {
        if (isInCheck(board, board.currentPlayer().pieceColor())) {
            return cannotMoveWithoutBeingCheck(board);
        } else {
            return false;
        }
    }

    public static boolean isADraw(BoardReader board) {
        if (!isInCheck(board, board.currentPlayer().pieceColor()) && cannotMoveWithoutBeingCheck(board)) {
            return true;
        }

        return board.drawCounter() == 50;
    }

    public static boolean moveDoesNotCreateCheck(BoardReader board, Move move) {
        return !board.playAndUndo(move, () -> isInCheck(board, move.fromPiece().color()));
    }

    public static boolean moveDoesNotCreateCheck(BoardReader board, Square from, Square to) {
        return moveDoesNotCreateCheck(board, new Move(board.pieceAt(from), from, to));
    }

    public static boolean noMoreMovesAllowed(BoardReader board) {
        if (isCheckmate(board)) {
            logger.info("checkmate winner is " + board.oppositePlayer().name());
            return true;
        } else if (isADraw(board)) {
            logger.info("draw");
            return true;
        } else
            return false;
    }

    public static Stream<Move> generateMoves(BoardReader board, Square from) {
        return board.pieceAt(from).generateMoves(board, from);
    }

    public static Stream<Move> generateMoves(BoardReader board, Square from, Predicate<Move> predicate) {
        return generateMoves(board, from).filter(predicate);
    }

    public static Collection<Move> generateMoves(BoardReader board) {
        return Square.allSquares.stream().flatMap((from) -> {
            if (board.isPieceOfCurrentPlayer(board.pieceAt(from))) {
                return generateMoves(board, from, (move) -> moveDoesNotCreateCheck(board, move));
            }
            return Stream.empty();
        }).collect(Collectors.toList());
    }
}

File: app/src/main/kotlin/yagoc/board/BoardRules.kt
package yagoc.board

import yagoc.Yagoc
import yagoc.pieces.PieceColor
import yagoc.pieces.PieceType
import java.util.function.Predicate
import java.util.stream.Stream
import kotlin.streams.toList

object BoardRules {
    @JvmStatic
    fun isCorrectMove(board: BoardView, from: Square, to: Square): Boolean {
        return isCorrectMove(board, Move(board.pieceAt(from), from, to))
    }

    @JvmStatic
    fun isCorrectMove(board: BoardView, move: Move): Boolean {
        return move.fromPiece().isCorrectMove(board, move)
    }

    @JvmStatic
    fun cannotMoveWithoutBeingCheck(board: BoardView): Boolean {
        return generateMoves(board).stream().noneMatch { move: Move -> moveDoesNotCreateCheck(board, move) }
    }

    @JvmStatic
    fun isInCheck(board: BoardView, color: PieceColor): Boolean {
        // Could not find $color king!
        val kingSquare = allSquares.first { square: Square ->
            board.pieceAt(square).pieceType == PieceType.King && board.pieceAt(square).color == color
        }

        return allSquares
            .any { from: Square ->
                board.someAt(from) && board.pieceAt(from).color != color && isCorrectMove(
                    board,
                    from,
                    kingSquare
                )
            }
    }

    fun isCurrentPlayerCheckmate(board: BoardView) =
        if (isInCheck(board, board.currentPlayer().pieceColor)) {
            cannotMoveWithoutBeingCheck(board)
        } else {
            false
        }

    fun isADraw(board: BoardView) =
        if (!isInCheck(board, board.currentPlayer().pieceColor) && cannotMoveWithoutBeingCheck(board)) {
            true
        } else {
            board.drawCounter() == 50
        }

    @JvmStatic
    fun moveDoesNotCreateCheck(board: BoardView, move: Move): Boolean {
        return !board.playAndUndo(move) { isInCheck(board, move.fromPiece().color) }
    }

    @JvmStatic
    fun moveDoesNotCreateCheck(board: BoardView, from: Square, to: Square): Boolean {
        return moveDoesNotCreateCheck(board, Move(board.pieceAt(from), from, to))
    }

    @JvmStatic
    fun noMoreMovesAllowed(board: BoardView): Boolean {
        return when {
            isCurrentPlayerCheckmate(board) -> true.also {
                Yagoc.logger.info("checkmate winner is " + board.oppositePlayer().name)
            }
            isADraw(board) -> true.also {
                Yagoc.logger.info("draw")
            }
            else -> false
        }
    }

    @JvmStatic
    fun generateMoves(board: BoardView, from: Square): Stream<Move> {
        return board.pieceAt(from).generateMoves(board, from)
    }

    @JvmStatic
    fun generateMoves(board: BoardView, from: Square, predicate: Predicate<Move>): Stream<Move> {
        return generateMoves(board, from).filter(predicate)
    }

    @JvmStatic
    fun generateMoves(board: BoardView): Collection<Move> {
        return allSquares.flatMap { from: Square ->
            if (board.isPieceOfCurrentPlayer(board.pieceAt(from))) {
                generateMoves(board, from) { move: Move -> moveDoesNotCreateCheck(board, move) }.toList()
            } else {
                emptyList()
            }
        }
    }
}